### DISCUSSION TITLE
Shaking while performing slam with NAO

### DISCUSSION THREAD

**Discussion by ysys98** [ros-naoqi/naoqi_driver#151] [Q&A] [answered]
Dear mbusy
when I set a 2D nav goal at rviz ,robot will move towards the target
but a strange phenomenon ：If the target point I set is approximately a straight line, the robot will walk towards the target point, but if the target point is in the oblique front of the robot, the robot will also walk but will not turn，the robot will not follow the globally planned path,robot just would walk sideways like a crab and will not avoid obstacles

What should I do？

Best wish

**Comment by mbusy** [comment]
Hi @ysys98, I see that you closed the issue, did you solve your problem?

**Comment by ysys98** [comment]
> Hi @ysys98, I see that you closed the issue, did you solve your problem?

@mbusy Hi, thank you for your reply. The previous problem has been solved. I still have some problems now. When I set a 2D NAV goal for Nao, he will walk along the global path and local path, but there will be a strange phenomenon that Nao swings left and right in the process of walking, because my obstacle update completely depends on the camera above Nao, Swinging left and right will cause great errors in the location of new obstacles in the local cost map. How can I avoid the problem of Nao swinging left and right during navigation？

**Comment by mbusy** [comment]
I'm afraid that it's a purely NAOqi related issue, if I understand correctly NAO's walk is responsible for the shaking of the camera? I guess you could use a different walk engine to generation joint trajectories allowing the robot to walk differently, but ultimately I think that you will still obtain some sort of shaking

**Comment by ysys98** [comment]
> I'm afraid that it's a purely NAOqi related issue, if I understand correctly NAO's walk is responsible for the shaking of the camera? I guess you could use a different walk engine to generation joint trajectories allowing the robot to walk differently, but ultimately I think that you will still obtain some sort of shaking

@mbusy Thank you for your help. Nao must shake when walking, causing the camera to shake slightly, which will not affect the accuracy of obstacles. In my experiment, the reason why Nao shakes left and right during navigation is because of the configuration problem of move_base caused Nao to swing left and right like a snake while walking towards the target point（2D nav goal）. I'm not sure which parameters are not configured well, which leads to this strange phenomenon

**Comment by mbusy** [comment]
Ok got it, the "natural shaking" when walking is fine, but the one you're experiencing is too intense. Do you have elements that tie that extra shaking to the configuration problem of move_base? (I unfortunately don't have a NAO to perform tests, it might be hard to correct that)

**Comment by ysys98** [comment]
> Ok got it, the "natural shaking" when walking is fine, but the one you're experiencing is too intense. Do you have elements that tie that extra shaking to the configuration problem of move_base? (I unfortunately don't have a NAO to perform tests, it might be hard to correct that)

@mbusy I think it's related to my local planner param configuration. I'm a novice in this field, so I don't know which parameters are related. Here's my move_ base parameter configuration. and when I use Nao to build a slam map, because the odometer of Nao itself will cause the coordinate error of the map to increase with the increase of the map area, how can I avoid this problem？By the way, I have always wondered whether NAO is a differential robot or an omni-directional robot？

`here is costmap_common_param.yaml
`
max_obstacle_height: 0.60  # assume something like an arm is mounted on top of the robot

footprint: [[ 0.3,  0.3], [-0.3,  0.3], [-0.3, -0.3], [ 0.3, -0.3]]


map_type: voxel

obstacle_layer:
  enabled:              true
  max_obstacle_height:  0.6
  origin_z:             0.0
  z_resolution:         0.2
  z_voxels:             2
  unknown_threshold:    15
  mark_threshold:       0
  combination_method:   1
  track_unknown_space:  true    #true needed for disabling global path planning through unknown space
  obstacle_range: 2.5
  raytrace_range: 3.0
  origin_z: 0.0
  z_resolution: 0.2
  z_voxels: 2
  publish_voxel_map: false
  observation_sources:  scan bump
  scan:
    data_type: LaserScan
    topic: scan
    marking: true
    clearing: true
    min_obstacle_height: 0.25
    max_obstacle_height: 0.35
  bump:
    data_type: PointCloud2
    topic: mobile_base/sensors/bumper_pointcloud
    marking: true
    clearing: false
    min_obstacle_height: 0.0
    max_obstacle_height: 0.15

inflation_layer:
  enabled:              true
  cost_scaling_factor:  5.0  # exponential rate at which the obstacle cost drops off (default: 10)
  inflation_radius:     0.5  # max. distance from an obstacle at which costs are incurred for planning paths.

static_layer:
  enabled:              true
`
here is global_costmap_param.yaml
`
global_costmap:
   global_frame: /odom
   robot_base_frame: /base_link
   update_frequency: 1.0
   publish_frequency: 0.5
   static_map: true
   transform_tolerance: 5
   plugins:
     - {name: static_layer,            type: "costmap_2d::StaticLayer"}
     - {name: obstacle_layer,          type: "costmap_2d::VoxelLayer"}
     - {name: inflation_layer,         type: "costmap_2d::InflationLayer"}

`
here is local_costmap_param.yaml
`

local_costmap:
   global_frame: odom
   robot_base_frame: /base_link
   update_frequency: 5.0
   publish_frequency: 2.0
   static_map: false
   rolling_window: true
   width: 4.0
   height: 4.0
   resolution: 0.05
   transform_tolerance: 5
   plugins:
    - {name: obstacle_layer,      type: "costmap_2d::VoxelLayer"}
    - {name: inflation_layer,     type: "costmap_2d::InflationLayer"}




`

`

**Comment by mbusy** [comment, marked as answer]
Since the issue is probably not related to the driver in itself, but is rather a naoqi / mapping problem, I'll convert the issue to a discussion.

I still don't really understand why the planner would cause the robot to "shake", but then I guess that if it sends contradictory move commands at high speed you might eventually experience some shaking.

> when I use Nao to build a slam map, because the odometer of Nao itself will cause the coordinate error of the map to increase with the increase of the map area, how can I avoid this problem？

I won't be able to give you a lot of information on mapping, but if only the odometry is used to localize the robot, its positioning error will increase with time, that's pretty much unavoidable. Using the cameras (visual slam) or a combination of vision and odometry (visual odometric slam) will avoid that ever inflating error problem, you already tested that I guess

> By the way, I have always wondered whether NAO is a differential robot or an omni-directional robot?

NAO is a legged robot, its walk engine allows him to walk in any direction, making it omni directional (this [rather old paper](https://ieeexplore.ieee.org/document/5686291) should give you more information)

**Reply by mbusy** [reply]
Additionally, could you make sure that only one process sends move commands to the robot? If multiple processes send contradictory move commands, the result might very well be a shaky behaviour

**Reply by ysys98** [reply]
@mbusy Thank you for your help， After I reduced the angular velocity, the shaking of the robot was effectively controlled，and using the cameras (visual slam) or a combination of vision and odometry (visual odometric slam) actually avoid that ever inflating error problem

**Reply by mbusy** [reply]
@ysys98 Great, glad to heat that! (You should mark as answer the comment that helped you, for the people using that discussion in the future :wink:)

**Reply by ysys98** [reply]
@mbusy of course! Can you tell me what I should do mark as answer the comment that helped me? I'm not very skilled for github ，thank you！

**Reply by mbusy** [reply]
@ysys98 Sure! But it seems that you figured it out on your own, the comment has been marked as an answer on my end :smile:
